### PR TITLE
Support reinterpretation for failing commands

### PR DESCRIPTION
**Pull request by nickchapman-da**
Support reinterpretation for failing commands

- `Engine.reinterpret` calls new compile method
- `Compiler.compileCommandForReinterpretation`
- which wraps `catchEverything` around the compiled command, so a rollback node is produced.
- test that uncatchable errors remains failures

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:1528]:
```suggestion
  private[this] val handleEverything: SExpr = SBSome(SEUpdatePureUnit)
```

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala
Comment (remyhaemmerle-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:144]: Maybe we can extend the comment to explain what's happening when the command fails with an unhanded exception ?

**Comment by remyhaemmerle-da** [comment]
cc @oliverse-da @andreaslochbihler-da

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ReinterpretTest.scala
Comment (cocreature) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ReinterpretTest.scala:26]: Can we get some tests for the edge cases here, I’m particularly interested in uncatchable errors due to
1. version checks
2. because they are uncatchable in general, e.g. failed DAML-LF ensure, error call, ….

**Review by cocreature — APPROVED** [review]
lgtm provided we extend the tests before merging, thanks!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala]: done

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala
Comment (nickchapman-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:148]: done

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ReinterpretTest.scala
Comment (nickchapman-da) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ReinterpretTest.scala:31]: failure ensure -- done
version check -- still to do
error call -- do we have an _uncatchable_ error primitive?

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ReinterpretTest.scala
Comment (cocreature) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ReinterpretTest.scala:151]: This is not the case I was trying to get at. I mean the following:
1. Exercise C
2. In that exercise create a contract with an ensure that returns `False` (so not one produced by the compiler, those throw so they are catchable).

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ReinterpretTest.scala
Comment (cocreature) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ReinterpretTest.scala:31]: Yes, Daml-LF’s error primitive is uncatchable. Daml’s `error` function does not use that primitive so you have to hand-write that to test it.
I think testing one uncatchable error is sufficient so either failed ensure (see my comment below, I think your test isn’t quite testing the right case), error call, ordering failure, ….

Comment (nickchapman-da) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ReinterpretTest.scala:31]: Removed the previous useless ensure test.
And added a test which provokes an uncatchable error by creating a broken `Ord` instance (as we discussed),

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ReinterpretTest.scala
Comment (nickchapman-da) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ReinterpretTest.scala]: didn't do this, as we discussed

**Comment by nickchapman-da** [comment]
Added tests that uncatchable errors remain as failures for reinterpretation.

**Comment by nickchapman-da** [comment]
We are ready to merge now

**Review by cocreature — APPROVED** [review]
Thanks for the tests!

**Pull request merged**
